tensorboard/loader.py

```diff
@@ -0,0 +1,414 @@
+# Copyright 2017 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""TensorBoard data ingestion module."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import collections
+import re
+import time
+import threading
+
+import six
+import tensorflow as tf
+
+from tensorboard import util
+
+_SHORTEN_EVENT_LOG_PATH_PATTERN = re.compile(r'(?:[^/\\]+[/\\])?(?:[^/\\]+)$')
+
+
+class Record(collections.namedtuple('Record', ('record', 'offset'))):
+  """Value class for a record returned by RecordReader.
+
+  Fields:
+    record: The byte string record that was read.
+    offset: The byte offset in the file *after* this record was read.
+
+  :type record: str
+  :type offset: int
+  """
+  __slots__ = ()  # Enforces use of only tuple fields.
+
+
+@util.closeable
+@six.python_2_unicode_compatible
+class RecordReader(object):
+  """Pythonic veneer around PyRecordReader."""
+
+  def __init__(self, path, start_offset=0):
+    """Creates new instance.
+
+    Args:
+      path: Path of file. This can be on a remote file system if the
+          TensorFlow build supports it.
+      start_offset: Byte offset to seek in file once it's opened.
+
+    :type path: str
+    :type start_offset: int
+    """
+    self.path = tf.compat.as_text(path)
+    self._offset = start_offset
+    self._reader = None  # type: tf.pywrap_tensorflow.PyRecordReader
+    self._is_closed = False
+
+  def get_next_record(self):
+    """Reads record from file.
+
+    Returns:
+      A Record or None if no more were available.
+
+    Raises:
+      IOError: On open or read error, or if close was called.
+
+    :rtype: Record
+    """
+    if self._is_closed:
+      raise IOError('%s is closed' % self)
+    if self._reader is None:
+      self._reader = self._open()
+    try:
+      with tf.errors.raise_exception_on_not_ok_status() as status:
+        self._reader.GetNext(status)
+    except (tf.errors.DataLossError, tf.errors.OutOfRangeError):
+      # We ignore partial read exceptions, because a record may be truncated.
```

Comment:
This means we are going to suppress these errors in the event where it failed for some reason other than the case you're planning for. In which case some other failure may be masked by returning None, which is standard behavior if the file is finished.

If retrying will succeed in this case, instead of returning None, maybe:
- make get_next_record a wrapper for _get_next_record(self, is_retry=False)
- if is_retry=True and it gets these errors, it re-raises them rather than returning None